src/audio/selector.c

```diff
@@ -0,0 +1,448 @@
+	case COMP_CMD_GET_VALUE:
+		return selector_ctrl_get_cmd(dev, cdata, max_data_size);
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * \brief Sets component component state.
+ * \param[in,out] dev Selector base component device.
+ * \param[in] cmd Command type.
+ * \return Error code.
+ */
+static int selector_trigger(struct comp_dev *dev, int cmd)
+{
+	trace_selector("selector_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+
+/**
+ * \brief Copies and processes stream data.
+ * \param[in,out] dev Selector base component device.
+ * \return Error code.
+ */
+static int selector_copy(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+
+	tracev_selector("selector_copy()");
+
+	/* selector components will only ever have 1 source and 1 sink buffer */
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+
+	/* make sure source component buffer has enough data available and that
+	 * the sink component buffer has enough free bytes for copy. Also
+	 * check for XRUNs
+	 */
+	if (source->avail < cd->source_period_bytes) {
+		trace_selector_error("selector_copy() error: "
+					"source component buffer"
+					" has not enough data available");
+		comp_underrun(dev, source, cd->source_period_bytes, 0);
+		return -EIO;	/* xrun */
+	}
+	if (sink->free < cd->sink_period_bytes) {
+		trace_selector_error("selector_copy() error: "
+						"sink component buffer"
+						" has not enough free bytes for copy");
+		comp_overrun(dev, sink, cd->sink_period_bytes, 0);
+		return -EIO;	/* xrun */
+	}
+
+	/* copy selected channels from in to out */
+	cd->sel_func(dev, sink, source);
+
+	/* calc new free and available */
+	comp_update_buffer_produce(sink, cd->sink_period_bytes);
+	comp_update_buffer_consume(source, cd->source_period_bytes);
+
+	return dev->frames;
+}
+
+/**
+ * \brief Prepares selector component for processing.
+ * \param[in,out] dev Selector base component device.
+ * \return Error code.
+ */
+static int selector_prepare(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sinkb;
+	struct comp_buffer *sourceb;
+	struct sof_ipc_comp_config *config = COMP_GET_CONFIG(dev);
+	int ret;
+
+	trace_selector("selector_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	/* selector components will have 1 source and 1 sink buffer */
+	sourceb = list_first_item(&dev->bsource_list,
+				  struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list,
+				struct comp_buffer, source_list);
+
+	/* get source data format */
+	comp_set_period_bytes(sourceb->source, dev->frames, &cd->source_format,
+			      &cd->source_period_bytes);
+
+	/* get sink data format */
+	comp_set_period_bytes(sinkb->sink, dev->frames, &cd->sink_format,
+			      &cd->sink_period_bytes);
+
```

Comment:
What are these 2 comp_set_period_bytes() calling really want to do here?
1. They are commented as "get xxx format" but actually SET period bytes and format.
2. Does it really work to set the sink component to work in 1 channel mode? The comp_set_period_bytes() invoking actually just use dev->params.channels of the sinkb->sink component directlly?

Reply:
For capture pipelines params will go upstream, so it should work correctly.
Added rewriting number of channels in selector_params to correctly propagate them further.